### PR TITLE
Clustering/base classes

### PR DESCRIPTION
**Pull request by MischaPanch**
Hello guys!

I'm finally back to sensAI after a break :). Here is the first of a series of PRs that are concerned with clustering. I want to include the new features step by step so you are not overwhelmed by the amount of code and can review effectively. The PRs following this one will contain the evaluation and ground truth modules as well as more features for dealing with geospatial clustering.

I have included a notebook that gives an overview of the most important features and interfaces so far. Here is the sample data needed to run it. 
[data.zip](https://github.com/jambit/sensAI/files/4972034/data.zip)

If we decide to put the sample data into git, e.g. through git lfs, I would add unit and integration tests

### CODE REVIEW

**Comment by MischaPanch** [comment]
Here the data needed to run the notebook

[data.zip](https://github.com/jambit/sensAI/files/4972027/data.zip)

**Comment by opcode81** [comment]
Please use cache.dumpPickle/loadPickle in your serialisation mixin

**Comment by MischaPanch** [comment]
> Please use cache.dumpPickle/loadPickle in your serialisation mixin

Done. I have also based VectorModel on the mixin. We talked about renaming util.cache to util.persistence but I haven't done it yet, as it is a breaking change. If we want to do that. I'd rename it in one of the PRs to follow

File: src/sensai/clustering/base/clustering.py [hunk 1]
```diff
@@ -0,0 +1,196 @@
...
+        if clusterId not in self.getLabels():
+            raise KeyError(f"no cluster for id {clusterId}")
+        return self.Cluster(self.datapoints[self.getLabels() == clusterId], identifier=clusterId)
+
+    @property
+    def numClusters(self) -> int:
+        return len(self.clusterIdentifiers.difference({self.noiseLabel}))
+
+    @abstractmethod
+    def _fit(self, x: np.ndarray) -> None:
```
Comment (opcode81) [inline src/sensai/clustering/base/clustering.py:154]: This is quite a strange interface. The purpose of this function seems to be to produce a side-effect that achieves that labels can subsequently be retrieved via _getLabels. Wouldn't a more fitting interface (pun intended) make the function _fit return whatever _getLabels returns?
We could give the function a clearer name, too (_computeLabels).

Comment (opcode81) [reply src/sensai/clustering/base/clustering.py:154]: Prefer functionally defined interfaces whenever possible.

File: src/sensai/base/interfaces.py [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+from typing import TypeVar
+
+import geopandas as gp
+
+T = TypeVar("T")
+
+
+# TODO: move this once more clustering-related stuff has been merged
+# Not a pure interface but this is still an appropriate place for it as the default plot method will often be overrode
+class GeoDataFrameWrapper(ABC):
```
[the remark this answers is not in the document]
Comment (MischaPanch) [reply src/sensai/base/interfaces.py:15]: Agree, see the TODO above. I would move it once we have more clustering stuff merged and the structure for it is finalized

File: src/sensai/util/cache.py [hunk 3]
```diff
@@ -576,8 +578,45 @@ def __init__(self, cacheBasePath: str, filenamePrefix: str = None, filename: str
...
+    def save(self, path: str) -> None:
+        pass
+
+    @classmethod
+    @abstractmethod
+    def load(cls: T, path: str) -> T:
+        pass
+
+
+class PickleSerializingMixin(LoadSaveInterface):
```
[the remark this answers is not in the document]
Comment (MischaPanch) [reply src/sensai/util/cache.py:599]: Well, it's using pickle in particular. I have another class that implements the same interface with geopandas (saving to and loading from a shapefile)

File: src/sensai/clustering/base/clustering.py (hunk 1 above)
Comment (MischaPanch) [reply src/sensai/clustering/base/clustering.py:154]: I wrote the class around the scikit-learn clustering algorithms which works exactly in the way you described, this made the SKLearnClusteringModel implementation very natural. But I agree, it will makes things easier to let _fit return the labels. I'll adjust it

File: src/sensai/util/cache.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (opcode81) [reply src/sensai/util/cache.py:599]: OK, I assumed there would (probably) only ever be pickle. In that case, how about PickleLoadSaveMixin?
I don't like "Serializing" because (a) it uses American spelling and (b) it is not as specific with regard to which functions are being mixed in (and strictly speaking does not even mention deserialisation). But that's just me.

File: src/sensai/base/interfaces.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (opcode81) [reply src/sensai/base/interfaces.py:15]: Haha, OK, missed that TODO :)

File: src/sensai/clustering/base/clustering.py [hunk 4]
```diff
@@ -0,0 +1,196 @@
...
+
+        def __len__(self):
+            return len(self.datapoints)
+
+        def __str__(self):
+            return f"{self.__class__.__name__}_{self.identifier}"
+
+        def __eq__(self, other):
+            if isinstance(other, self.__class__):
+                return self.identifier == other.identifier and np.array_equal(self.datapoints, other.datapoints)
```
[the remark this answers is not in the document]
Comment (MischaPanch) [reply src/sensai/clustering/base/clustering.py:50]: It could happen when clusters are loaded from different files, therefore the check

File: src/sensai/clustering/base/clustering.py (hunk 1 above)
Comment (MischaPanch) [reply src/sensai/clustering/base/clustering.py:154]: Done, thanks for the comment! It's much cleaner now

File: src/sensai/clustering/base/clustering.py [hunk 5]
```diff
@@ -0,0 +1,196 @@
...
+            return np.mean(self.datapoints, axis=0)
+
+        def radius(self):
+            return np.max(distance_matrix([self.centroid()], self.datapoints))
+
+        def summaryDict(self):
+            """
+            Dictionary containing coarse information about the cluster (e.g. num_members and centroid)
+
+            :return: A dict
```
[the remark this answers is not in the document]
Comment (MischaPanch) [reply src/sensai/clustering/base/clustering.py:63]: Done (hopefully everywhere)

File: src/sensai/clustering/greedy_clustering.py [hunk 6]
```diff
@@ -5,8 +5,9 @@
 from abc import ABC, abstractmethod
 
 log = logging.getLogger(__name__)
-         
 
+
+# TODO: inherit from ClusteringModel once we have agreed on the interface
```
[the remark this answers is not in the document]
Comment (MischaPanch) [reply src/sensai/clustering/greedy_clustering.py:10]: Done

File: src/sensai/util/cache.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (MischaPanch) [reply src/sensai/util/cache.py:599]: Done

File: src/sensai/util/tracking.py [hunk 7]
```diff
@@ -0,0 +1,18 @@
+import logging
+import time
+from typing import Callable
+
+log = logging.getLogger(__name__)
+
+
+def timeit(method: Callable):
```
[the remark this answers is not in the document]
Comment (MischaPanch) [reply src/sensai/util/tracking.py:8]: Renamed it

File: src/sensai/clustering/base/clustering.py [hunk 8]
```diff
@@ -0,0 +1,196 @@
...
+
+    @abstractmethod
+    def _getLabels(self) -> np.ndarray:
+        """
+        :return: list of the same length as the input datapoints; it represents the mapping coordinate -> cluster_id
+        """
+        pass
+
+
+class SKLearnTypeClusterer(Protocol):
```
[the remark this answers is not in the document]
Comment (MischaPanch) [reply src/sensai/clustering/base/clustering.py:165]: Done

File: src/sensai/clustering/base/clustering.py [hunk 9]
```diff
@@ -0,0 +1,196 @@
...
+
+        if minClusterSize is not None or maxClusterSize is not None:
+            if noiseLabel is None:
+                raise ValueError("the noise label has to be not None for non-trivial bounds on cluster sizes")
+        self.noiseLabel = noiseLabel
+        self.maxClusterSize = maxClusterSize if maxClusterSize is not None else np.inf
+        self.minClusterSize = minClusterSize if minClusterSize is not None else -np.inf
+
+    class Cluster:
+        def __init__(self, datapoints: np.ndarray, identifier: Union[int, str]):
```
[the remark this answers is not in the document]
Comment (MischaPanch) [reply src/sensai/clustering/base/clustering.py:38]: I adjusted the todo above the class hinting that we might want to do that in the future

File: src/sensai/clustering/base/clustering.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (opcode81) [reply src/sensai/clustering/base/clustering.py:50]: This is an equivalence relation though, not a check for inconsistent clusters. If the identifier doesn't actually identify the cluster, then there's a problem!

[the remark this answers is not in the document]
Comment (opcode81) [reply src/sensai/clustering/base/clustering.py:50]: Two questions come to mind: 
(1) Why do we need the equivalence relation in its current state?
(2) Why do we need an identifier that doesn't actually identify the cluster? (We should get rid of it.)

[the remark this answers is not in the document]
Comment (opcode81) [reply src/sensai/clustering/base/clustering.py:50]: I guess the answer is that you want to compare clusters that are associated with different clustering processes?

File: src/sensai/clustering/base/clustering.py [hunk 10]
```diff
@@ -0,0 +1,196 @@
...
+    # unfortunately, there seems to be no way to annotate the return type correctly
+    # https://github.com/python/mypy/issues/3993
+    def getCluster(self, clusterId: int) -> Cluster:
+        if clusterId not in self.getLabels():
+            raise KeyError(f"no cluster for id {clusterId}")
+        return self.Cluster(self.datapoints[self.getLabels() == clusterId], identifier=clusterId)
+
+    @property
+    def numClusters(self) -> int:
+        return len(self.clusterIdentifiers.difference({self.noiseLabel}))
```
Comment (opcode81) [inline src/sensai/clustering/base/clustering.py:148]: Since we use this difference construct several times, we might want to consider leaving the noise label out or keeping two lists of identifiers. Especially since you used the property annotation here, I expect this to be O(1), but that's far from the case. I highly discourage the use of the property annotation particularly in such situations. (Basically, I think it should only be used as a tool to establish backward compatibility - when a property has been removed but can be functionally determined by other properties - and not as a general-purpose pattern.)

_General observation_: Your implementations don't put much emphasis on computational efficiency. We can certainly apply such a style in our own project-specific code, but we should pay more attention to efficiency in the library (whenever it doesn't lead to overly complex code), as inefficiencies could lead to severe runtime penalties down the line.

File: src/sensai/clustering/base/clustering.py [hunk 11]
```diff
@@ -0,0 +1,196 @@
...
+        return labels
+
+    @property
+    def clusterIdentifiers(self) -> Set[int]:
+        assert self.isFitted
+        return self._clusterIdentifiers
+
+    # unfortunately, there seems to be no way to annotate the return type correctly
+    # https://github.com/python/mypy/issues/3993
+    def getCluster(self, clusterId: int) -> Cluster:
```
[the remark this answers is not in the document]
Comment (opcode81) [reply src/sensai/clustering/base/clustering.py:141]: I should not fail to mention that methods in eval_stats are also guilty of this - and I have considered changing that more than once, but backward compatibility seemed more important in the past, but we should improve the names at some point.

File: src/sensai/clustering/base/clustering.py (hunk 10 above)
Comment (MischaPanch) [reply src/sensai/clustering/base/clustering.py:148]: This one I can get in O(1) simply by substracting 1 when the noise label is not None. That should be good enough, right?

File: src/sensai/clustering/base/clustering.py [hunk 12]
```diff
@@ -0,0 +1,196 @@
...
+    @property
+    def isFitted(self):
+        return self._datapoints is not None
+
+    @property
+    def datapoints(self) -> np.ndarray:
+        assert self.isFitted
+        return self._datapoints
+
+    def getLabels(self) -> np.ndarray:
```
[the remark this answers is not in the document]
Comment (MischaPanch) [reply src/sensai/clustering/base/clustering.py]: Labels are cached in the _labels field and this is just a getter. I'm not sure I understand your point

File: src/sensai/clustering/base/clustering.py [hunk 13]
```diff
@@ -0,0 +1,196 @@
...
+
+        def __str__(self):
+            return f"{self.__class__.__name__}_{self.identifier}"
+
+        def __eq__(self, other):
+            if isinstance(other, self.__class__):
+                return self.identifier == other.identifier and np.array_equal(self.datapoints, other.datapoints)
+            return False
+
+        def centroid(self):
```
[the remark this answers is not in the document]
Comment (MischaPanch) [reply src/sensai/clustering/base/clustering.py]: Done, in a verbose but extendible way

File: src/sensai/clustering/base/clustering.py [hunk 14]
```diff
@@ -0,0 +1,196 @@
...
+        labels = self._getLabels()
+        # Relabel clusters that do not fulfill size bounds as noise
+        if self.minClusterSize != -np.inf or self.maxClusterSize != np.inf:
+            for clusterId, clusterSize in zip(*np.unique(labels, return_counts=True)):
+                if not self.minClusterSize <= clusterSize <= self.maxClusterSize:
+                    labels[labels == clusterId] = self.noiseLabel
+        return labels
+
+    @property
+    def clusterIdentifiers(self) -> Set[int]:
```
[the remark this answers is not in the document]
Comment (MischaPanch) [reply src/sensai/clustering/base/clustering.py:150]: I added a type hint and enhanced the docstring concerned with labels
